services/QuillLMS/engines/evidence/lib/evidence/evidence/open_ai.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+module Evidence
+  class OpenAI < Evidence::ApplicationService
+    include HTTParty
+    base_uri 'https://api.openai.com/v1'
+
+    class NoResultsError < StandardError; end
+
+    API_KEY = ENV['OPENAI_API_KEY']
+    MAX_TOKENS = 24
+    COMPLETION_ENDPOINT = '/completions'
+    # A-D, (A)Least -> (D)Most Complex/Expensive
+    COMPLETION_MODELS = {
+      ada: 'text-ada-001',
+      babbage: 'text-babbage-001',
+      curie: 'text-curie-001',
+      davinci: 'text-davinci-002'
+    }
+    BLANK = ''
+    STOP_TOKENS = [". ", ", "]
+
+    attr_accessor :response, :prompt, :temperature, :count, :model_key, :options_hash
+
+    def initialize(prompt:, temperature: 0.5, count: 1, model_key: :babbage, options_hash: {})
+      @prompt = prompt
+      @temperature = temperature
+      @count = count
+      @model_key = model_key
+      @options_hash = options_hash
+    end
+
+    def headers
+      {
+        "Content-Type" => "application/json",
+        "Authorization" => "Bearer #{API_KEY}"
+      }
+    end
+
+    def request_body
+      {
+        model: COMPLETION_MODELS[model_key],
+        temperature: temperature,
+        prompt: prompt,
+        n: count,
+        max_tokens: MAX_TOKENS,
+        stop: STOP_TOKENS
+      }.merge(options_hash)
+    end
+
+    private def result_texts
+      response
+        .parsed_response['choices']
+        .map{|r| r['text']}
+    end
+
+    def cleaned_results
+      result_texts
+        .map{|r| r.strip } # remove leading/ending spaces
+        .map{|r| r.gsub(/^(\n)+/, BLANK)} # strip all leading \n
+        .map{|r| r.gsub(/^-/, BLANK)} # strip leading dash
+        .map{|r| r.gsub(/\d\)/, BLANK)} # strip 1), 2), 3)
+        .map{|r| r.gsub(/(\]|\[|\=)/, BLANK)} # strip brackets and equal signs
+        .map{|r| r.split(/\n/).first } # drop anything after a \n
+        .map{|r| r.strip } # remove leading/ending spaces again
+        .compact
+        .select {|r| r.length >= 10}
+        .uniq
+    end
+
+    def run
+      return response if response.present?
+
+      @response = self.class.post(COMPLETION_ENDPOINT, body: request_body.to_json, headers: headers)
+
+      cleaned_results
+    end
+
+    def reset_response
+      @response = nil
+    end
+
+    # Edit endpoint code still in experimentation phase
+    EDITS_ENDPOINT = '/edits'
+    EDIT_MODELS = {
+      ada: 'text-ada-edit-001',
+      babbage: 'text-babbage-edit-001',
+      curie: 'text-curie-edit-001',
+      davinci: 'text-davinci-edit-001'
```

Comment:
Note, I'm going to remove the edit code for now (stash it in a branch) since it's still untested. 

I just started experimenting with it, and according to the docs the `edit` models are "different" than the `completion` ones (note the 'edit' in the name), but for some reason it doesn't document all of their names. The davinci one is the only one I got to work (that name was in an example), but I was guessing at the rest of the names for now.